### PR TITLE
Allow RobotState::setFromIK to work with subframes

### LINKED ISSUE #3072: RobotState::setFromIK is subframe unaware
### Description

`RobotState::setFromIK()` returns false when the `tip` frame passed is a subframe. This is due to the method [only checking if the frame is the name of an attached body](https://github.com/moveit/moveit2/blob/ad0431b095955c03f3b26521562fb445873b7b6c/moveit_core/robot_state/src/robot_state.cpp#L1669-L1695). This is in contrast to the ROS1 code, which [calls `getRigidlyConnectedParentLinkModel()`](https://github.com/moveit/moveit/blob/f1a607129a6b5183c9296b29e0d126d9b48e7ea8/moveit_core/robot_state/src/robot_state.cpp#L1710-L1733), which [is subframe aware](https://github.com/moveit/moveit/blob/f1a607129a6b5183c9296b29e0d126d9b48e7ea8/moveit_core/robot_state/src/robot_state.cpp#L833). The code is a bit different in ROS2, however, `RobotState` [already has a method for looking up frames with subframe awareness](https://github.com/moveit/moveit2/blob/ad0431b095955c03f3b26521562fb445873b7b6c/moveit_core/robot_state/src/robot_state.cpp#L1082-L1148). As such, I think the call to `hasAttachedBody()` should be replaced with `getFrameTransform()`

### ROS Distro

Humble

### OS and version

Ubuntu 22.04

### Source or binary build?

Source

### If binary, which release version?

_No response_

### If source, which branch?

N/A

### Which RMW are you using?

None

### Steps to Reproduce

1. Call `RobotState::setFromIK()` where the `tip` parameter is a subframe
2. The code (will error and return false)[https://github.com/moveit/moveit2/blob/ad0431b095955c03f3b26521562fb445873b7b6c/moveit_core/robot_state/src/robot_state.cpp#L1682-L1683]

### Expected behavior

The call should recognise that the tip is a subframe, return true, and set the robot state accordingly. This is consistent with ROS1 behaviour. 

### Actual behavior

The code prints an error "The following Pose Frame does not exist: <subframe>", and returns false.

### Backtrace or Console output

_No response_

### PR DESCRIPTION
**Pull request by TSNoble**
### Description

Fixes #3072

- Adds passing regression / sanity tests which verify that a `RobotState` can `setFromIK` with a collision object
- Adds failing tests demonstrating that `RobotState` cannot `setFromIK` with a subframe
- Fixes failing tests by modifying `RobotState::setFromIK` to consider subframes
- Adds some helper functions to the pilz tests for attaching objects / subframes and checking joint values

I tried adding the tests directly to the robot state tests, however, I was getting issues with no kinematic solver being associated with the planning group of the `OneRobot` fixture.

Pilz already had a test which was checking `setFromIK`, and appears to be setup with a kinematic solver, so I've added the tests there.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/moveit/moveit2/blob/main/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [x] Include a screenshot if changing a GUI (no GUI changes)
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by TSNoble** [comment]
@sjahr I believe this is ready for review (just the usual tutorial image failures)

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (sea-bass) [inline moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:139]: Thanks for the docstring! Can you fill these in?

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (sea-bass) [inline moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:148]: Same here re: filling in

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (sea-bass) [inline moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:156]: And here

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (sea-bass) [inline moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:206]: Tend to use `++i` instead of `i++` in loops

**Review by sea-bass — COMMENTED** [review]
This is great! Especially appreciate the tests.

Mostly small comments from my side.

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (rr-tom-noble) [reply moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:139]: Sure thing. I'm guessing the synrtax is just @param <name> <description> ?

Comment (rr-tom-noble) [reply moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:139]: Done :+1:

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (rr-tom-noble) [reply moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:148]: Done :+1:

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (rr-tom-noble) [reply moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp]: Done :+1:

File: moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp
Comment (rr-tom-noble) [reply moveit_planners/pilz_industrial_motion_planner/test/unit_tests/src/unittest_trajectory_functions.cpp:156]: Done :+1:

**Comment by rr-tom-noble** [comment]
Ah, also if this could be marked for backport that'd be much appreciated. I'll get on fixing those PRs ASAP

**Comment by sea-bass** [comment]
> @sea-bass Should be ready for review again.
> 
> Additionally, I'm always a little unsure of what types of changes require extending the tutorials & api notes. In theory the interface hasn't changed (the `tip` parameter still accepts a string and is therefore backwards-compatible, but the semantics are a little different in that it now accepts subframe names)

Thank you!

Your assessment is right; API wise, you're fine.

That said, i think subframes as a whole are not generally super well documented in the tutorials. So if you're a big user of this feature, you should in futire consider contribjting a tutorial! That would be great.

**Review by sea-bass — APPROVED** [review]
LGTM, will await a 2nd review from someone else as well.

**Comment by rr-tom-noble** [comment]
@sjahr Would you mind providing a second review when you have time?

**Review by sjahr — APPROVED** [review]
Looks good thanks!

**Pull request merged**
